scripts/gha/utils.py

```diff
@@ -89,15 +90,79 @@ def is_linux_os():
  return platform.system() == 'Linux'
 
 
-def get_vcpkg_triplet(arch):
-  """ Get vcpkg target triplet (platform definition).
+def check_vcpkg_triplet(triplet_name, arch, crt_linkage):
+  """Check if a vcpkg triplet exists that match our specification.
 
   Args:
     arch (str): Architecture (eg: 'x86', 'x64').
+    crt_linkage (str): Runtime linkage for MSVC (eg: 'dynamic', 'static')
+
+  Returns:
+    (bool): If the triplet is valid.
+  """
+  triplet_file_path = get_vcpkg_triplet_file_path(triplet_name)
+  if not os.path.exists(triplet_file_path):
+    return False
+
+  _arch = None
+  _crt_linkage = None
+
+  with open(triplet_file_path, 'r') as triplet_file:
+    for line in triplet_file.readlines():
+      # Eg: set(VCPKG_TARGET_ARCHITECTURE x86)\n
+      line = line.rstrip('\n').rstrip(')')
+      # Eg: set(VCPKG_TARGET_ARCHITECTURE x86
+      if not line.startswith('set('):
+        continue
+      # Eg: 'set(', 'VCPKG_TARGET_ARCHITECTURE x86'
+      variable = line.split('set(')[-1]
+      variable_name, variable_value = variable.split(' ')
+      if variable_name == 'VCPKG_TARGET_ARCHITECTURE':
+        _arch = variable_value
+      elif variable_name == 'VCPKG_CRT_LINKAGE':
+        _crt_linkage = variable_value
+
+  return (_arch == arch) and (_crt_linkage == crt_linkage)
+
+
+def create_vcpkg_triplet(arch, crt_linkage):
+  """Create a new triplet configuration.
+
+  Args:
+    arch (str): Architecture (eg: 'x86', 'x64').
+    crt_linkage (str): Runtime linkage for MSVC (eg: 'dynamic', 'static')
+
+  Returns:
+    (str): Triplet name
+      Eg: 'x86-windows-dynamic'
+  """
+  contents = [ 'VCPKG_TARGET_ARCHITECTURE {0}'.format(arch),
+               'VCPKG_CRT_LINKAGE {0}'.format(crt_linkage),
+               'VCPKG_LIBRARY_LINKAGE static']
+
+  contents = ['set({0})\n'.format(content) for content in contents]
+  if is_linux_os() or is_mac_os():
+    contents.append('\n')
+    contents.append('set(VCPKG_CMAKE_SYSTEM_NAME {0})\n'.format(platform.system()))
+
+  triplet_name = '{0}-{1}-{2}'.format(arch.lower(), platform.system().lower(), crt_linkage)
```

Comment:
There's a function in this script to generate a triplet name, is there a reason we're avoiding that?

Reply:
Hmmm... I agree this is a little confusing but this function is called from get_vcpkg_triplet() if that function fails to find a triplet file for our configuration. 

"vcpkg" naming of triplet files adds a little more confusion to this system. There are 2 sorts of linkages that can be specified - runtime and library. 
```
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
```
The names of the prebuilt triplet names do not account for both of these triplets. If the values are chosen defaults for the platform (eg: for linux and mac, runtime linkage is dynamic and library_linkage is static), then the triplet name is `<architecture>-<os>' (eg: x64-linux, x64-osx). For windows, we have an additional triplet "x64-windows-static" where the runtime and library linkage both are set to static. 

In my case, I am including the runtime linkage (crt_linkage) as part of the name because library linkage should always be "static" for us all the time on all platforms.